<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A User Guide to the patherit R-package • patherit</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">patherit</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/UserGuide.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>A User Guide to the patherit R-package</h1>
                        <h4 class="author">Venelin Mitov</h4>
            
          </div>

    
    
<div class="contents">
Here we show how to use <strong>patherit</strong> to estimate the broad-sense heritability, <span class="math inline">\(H^2\)</span>, of a numeric pathogen trait. <span class="math inline">\(H^2\)</span> is defined as the proportion of the trait variance attributable to genotypic variation in a population of hosts <span class="citation">(Mitov and Stadler 2016)</span>:
<span class="math display">\[\begin{equation}
H^2=\frac{\text{Var}(g)}{\text{Var}(z)}
\end{equation}\]</span>
<p>Thus, the heritability is a number in the interval [0,1]. The closer <span class="math inline">\(H^2\)</span> is to 1, the higher is the importance of the pathogen genotype versus the host immune system in the formation of the trait. The input data represents a transmission tree connecting a set of infected patients and the patient’s trait-values, measured at the moment of diagnosis. The <strong>patherit</strong> package evaluates phylogenetic comparative estimators, i.e. <span class="math inline">\(H_{\bar{t}}^2\)</span>, <span class="math inline">\(H_{\infty}^2\)</span> and <span class="math inline">\(H_e^2\)</span> from the PMM and POUMM methods, and pair-correlation estimators, i.e. <span class="math inline">\(r_A\)</span> from the ANOVA-CPP method <span class="citation">(Mitov and Stadler 2016)</span>. The main reason for using different estimators is the need to gain certainty in the estimates and to eliminate potential sources of bias, (e.g. within-host evolution and wrong model assumptions), which can obfuscate the true value of <span class="math inline">\(H^2\)</span> <span class="citation">(Mitov and Stadler 2016)</span>.</p>
<p>For the purpose of this tutorial, we will use the toy-model epidemic simulated in the <a href="https://venelin.github.io/toyepidemic/articles/UserGuide.html">User guide</a> of the <a href="https://venelin.github.io/toyepidemic/index.html">toyepidemic</a> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">"epid1.RData"</span>)</code></pre></div>
<div id="extracting-phylogenetic-and-phenotypic-data-from-the-simulated-epidemic" class="section level2">
<h2 class="hasAnchor">
<a href="#extracting-phylogenetic-and-phenotypic-data-from-the-simulated-epidemic" class="anchor"></a>Extracting phylogenetic and phenotypic data from the simulated epidemic</h2>
<p>The code-snippet below extracts the exact transmission tree connecting a set of diagnosed patients during a simulated epidemic outbreak.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tree &lt;-<span class="st"> </span>toyepidemic<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/toyepidemic/topics/extractTree">extractTree</a></span>(epidemic)

<span class="co"># a data.table containing the host-type (env), the infecting strain (gene) and the</span>
<span class="co"># host specific effect (e) for each infected patient. </span>
pop &lt;-<span class="st"> </span>toyepidemic<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/toyepidemic/topics/extractPop">extractPop</a></span>(epidemic, <span class="dt">ids=</span>tree<span class="op">$</span>tip.label)

<span class="co"># use the function calcValue from the toyepidemic package to extract the trait</span>
<span class="co"># value for each sampled patient.</span>
pop[, z<span class="op">:</span><span class="er">=</span>toyepidemic<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/toyepidemic/topics/calcValue">calcValue</a></span>(env, gene, e, <span class="dt">GEValues=</span>epidemic<span class="op">$</span>GEValues)]

<span class="co"># arrange the trait values in order of the patient-labels in the tree and store</span>
<span class="co"># them in an external vector.</span>
z &lt;-<span class="st"> </span>pop[<span class="kw">list</span>(tree<span class="op">$</span>tip.label), z]</code></pre></div>
</div>
<div id="visualizing-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#visualizing-the-data" class="anchor"></a>Visualizing the data</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(ape<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/ape/topics/ladderize">ladderize</a></span>(tree), <span class="dt">show.tip.label =</span> <span class="ot">FALSE</span>, <span class="dt">type=</span><span class="st">"fan"</span>, <span class="dt">no.margin =</span> <span class="ot">TRUE</span>)
colorFunc &lt;-<span class="st"> </span><span class="kw">colorRampPalette</span>(<span class="kw">c</span>(<span class="st">"blue"</span>, <span class="st">"red"</span>))
colors &lt;-<span class="st"> </span><span class="kw">colorFunc</span>(<span class="kw">length</span>(tree<span class="op">$</span>tip.label))
ape<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/ape/topics/nodelabels">tiplabels</a></span>(<span class="ot">NULL</span>, <span class="dt">tip=</span><span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(tree<span class="op">$</span>tip.label), <span class="dt">pch=</span><span class="dv">20</span>, <span class="dt">col=</span>colors[<span class="kw">order</span>(z)], <span class="dt">cex=</span><span class="fl">0.5</span>)</code></pre></div>
<p><img src="UserGuide_files/figure-html/unnamed-chunk-3-1.png" width="680"></p>
<p>Qualitatively, the heritability of the trait manifests itself in the similar color for closely related patients on the transmission tree. This is because closely related patients such as closest phylogenetic pairs (CPPs) tend to carry identical pathogen strains.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">patherit<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/patherit/topics/boxplotTraitAlongTree">boxplotTraitAlongTree</a></span>(z, tree, <span class="dt">nGroups =</span> <span class="dv">8</span>, <span class="dt">boxwex=</span><span class="dv">6</span>)</code></pre></div>
<p><img src="UserGuide_files/figure-html/unnamed-chunk-4-1.png" width="680"></p>
<p>The absence of a noticeable trend and the relative symmetry of the box-whiskers around the medians suggest that the trait distribution is approximately normal and stationary throughout the epidemic. Thus, we should expect that the POUMM estimates <span class="math inline">\(H_{\bar{t}}^2\)</span> and <span class="math inline">\(H_{\infty}^2\)</span> should be approximately equal.</p>
</div>
<div id="estimating-the-heritability-using-anova-cpp-poumm-and-the-pmm-methods-" class="section level2">
<h2 class="hasAnchor">
<a href="#estimating-the-heritability-using-anova-cpp-poumm-and-the-pmm-methods-" class="anchor"></a>Estimating the heritability using ANOVA-CPP, POUMM and the PMM methods.</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># start a parallel cluster for parallel execution of MCMC chains in the POUMM</span>
<span class="co"># and PMM fits (explained in the vignette for the POUMM package).</span>
cluster &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/parallel/topics/makeCluster">makeCluster</a></span>(parallel<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/parallel/topics/detectCores">detectCores</a></span>(<span class="dt">logical =</span> <span class="ot">FALSE</span>))
doParallel<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/doParallel/topics/registerDoParallel">registerDoParallel</a></span>(cluster)

H2Analysis &lt;-<span class="st"> </span>patherit<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/patherit/topics/estimateH2">estimateH2</a></span>(z, tree, 
                                   <span class="dt">methods =</span> <span class="kw">list</span>(
                                     <span class="dt">PP =</span> <span class="kw">list</span>(<span class="dt">bootstraps=</span><span class="dv">100</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span>), 
                                     <span class="dt">POUMM =</span> <span class="kw">list</span>(<span class="dt">nSamplesMCMC =</span> <span class="fl">5e5</span>, 
                                                  <span class="dt">verbose =</span> <span class="ot">FALSE</span>), 
                                     <span class="dt">PMM =</span> <span class="kw">list</span>(<span class="dt">nSamplesMCMC =</span> <span class="fl">5e5</span>, 
                                                <span class="dt">verbose =</span> <span class="ot">FALSE</span>)))</code></pre></div>
</div>
<div id="summarizing-the-estimates" class="section level2">
<h2 class="hasAnchor">
<a href="#summarizing-the-estimates" class="anchor"></a>Summarizing the estimates:</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(H2Analysis)[stat<span class="op">%in%</span><span class="kw">c</span>(<span class="st">"rA"</span>, <span class="st">"H2tMean"</span>, <span class="st">"H2e"</span>, <span class="st">"H2tInf"</span>)]</code></pre></div>
<pre><code>##     method tauQuantile    N   K    stat  MLE PostMean CI.lower CI.upper tauMean filter
##  1:     PP    V[0%,5%]   28  14      rA 0.67       NA     0.26     0.88     3.3    all
##  2:     PP   D[0%,10%]   56  28      rA 0.55       NA     0.24     0.76     4.4    all
##  3:     PP   O[0%,12%]   70  35      rA 0.55       NA     0.27     0.74     4.9    all
##  4:     PP  qu[0%,20%]  112  56      rA 0.63       NA     0.45     0.77     6.1    all
##  5:     PP   Q[0%,25%]  138  69      rA 0.57       NA     0.39     0.71     6.7    all
##  6:     PP   M[0%,50%]  278 139      rA 0.54       NA     0.41     0.65    10.7    all
##  7:     PP  A[0%,100%]  552 276      rA 0.43       NA     0.33     0.52    19.5    all
##  8:  POUMM          NA 1000  NA H2tMean 0.62     0.62     0.49     0.74      NA     NA
##  9:  POUMM          NA 1000  NA     H2e 0.61     0.60     0.47     0.73      NA     NA
## 10:  POUMM          NA 1000  NA  H2tInf 0.62     0.62     0.49     0.74      NA     NA
## 11:    PMM          NA 1000  NA H2tMean 0.70     0.69     0.56     0.81      NA     NA
## 12:    PMM          NA 1000  NA     H2e 0.35     0.34     0.25     0.44      NA     NA
## 13:    PMM          NA 1000  NA  H2tInf 1.00     1.00     1.00     1.00      NA     NA</code></pre>
</div>
<div id="generating-a-correlation-profile" class="section level2">
<h2 class="hasAnchor">
<a href="#generating-a-correlation-profile" class="anchor"></a>Generating a correlation profile</h2>
<p>It is useful to observe the correlation between phylogenetic pairs as a function of their phylogenetic distance. In the absence of convergent evolution towards the same pathogen, the sequence similarity between transmission partners is supposed to decrease with the time since the transmission events. This is reflected by a decay in the trait correlaiton between transmission partners and phylogenetic pairs, respectively. The correlation profile represents a suite of estimates of the intra-class correlation (<span class="math inline">\(r_A\)</span>) in a stratification of the phylogenetic patients by their phylogenetic distance. The patherit package allows to produce such a profile for the original data, and for data simulated under the ML fits of the POUMM and the PMM. This is done with the <code><a href="http://www.rdocumentation.org/packages/patherit/topics/corrProfile">patherit::corrProfile()</a></code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">corrPr &lt;-<span class="st"> </span>patherit<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/patherit/topics/corrProfile">corrProfile</a></span>(H2Analysis)</code></pre></div>
<p>Then, we use the function <code><a href="http://www.rdocumentation.org/packages/patherit/topics/plotCorrProfile">patherit::plotCorrProfile()</a></code> to plot the correlation values in different stratifications (medians, quartiles, quantiles, deciles, etc…):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">patherit<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/patherit/topics/plotCorrProfile">plotCorrProfile</a></span>(H2Analysis, corrPr)</code></pre></div>
<p><img src="UserGuide_files/figure-html/unnamed-chunk-9-1.png" width="652.8"> Notice that the PMM estimates remain high for all phylogenetic distances, while the true data and the simulations under the POUMM ML fit are showing a decline in the correlation between phylogenetic pairs with phylogenetic distance. This shows that the PMM cannot model the decay in correlaiton <span class="citation">(Mitov and Stadler 2016)</span>.</p>
</div>
<div id="comparing-to-the-true-value" class="section level2">
<h2 class="hasAnchor">
<a href="#comparing-to-the-true-value" class="anchor"></a>Comparing to the true value</h2>
<p>For the toy-model simulation, we know the true value of <span class="math inline">\(H^2\)</span>. This is given by the direct estimator - the coefficient of determination, <span class="math inline">\(R_{adj}^2\)</span>, calculated over a grouping of the host population by identical infecting strain:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/R2adj.html">R2adj</a></span>(<span class="dt">data=</span>pop, <span class="dt">activeOnly =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>## [1] 0.55</code></pre>
<p>We see that the true value falls well within the 95% CI of all estimators except, the estimators <span class="math inline">\(H_e^2\)</span> from the PMM method. This is due to the fact that the PMM has over-estimated the non-heritable variance <span class="math inline">\(\sigma_e^2\)</span> - another manifestation of the PMM’s inability to model the decay of correlation between phylogenetic pairs.</p>
</div>
<div id="randomization-test" class="section level2">
<h2 class="hasAnchor">
<a href="#randomization-test" class="anchor"></a>Randomization test</h2>
<p>The randomization test consists in checking that the estimated heritability is insignificant if the trait values have been shuffled randomly between the sampled patients:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pop[, z2<span class="op">:</span><span class="er">=</span><span class="kw">sample</span>(z)]
zR &lt;-<span class="st"> </span>pop[, z2]</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">H2AnalysisR &lt;-<span class="st"> </span><span class="kw"><a href="../reference/estimateH2.html">estimateH2</a></span>(zR, tree, 
                         <span class="dt">methods =</span> <span class="kw">list</span>(
                           <span class="dt">PP =</span> <span class="kw">list</span>(<span class="dt">bootstraps=</span><span class="dv">100</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span>), 
                           <span class="dt">POUMM =</span> <span class="kw">list</span>(<span class="dt">nSamplesMCMC =</span> <span class="fl">5e5</span>, 
                                        <span class="dt">verbose =</span> <span class="ot">FALSE</span>), 
                           <span class="dt">PMM =</span> <span class="kw">list</span>(<span class="dt">nSamplesMCMC =</span> <span class="fl">5e5</span>, 
                              <span class="dt">verbose =</span> <span class="ot">FALSE</span>)))

parallel<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/parallel/topics/makeCluster">stopCluster</a></span>(cluster)</code></pre></div>
<div id="summarizing-the-estimates-1" class="section level3">
<h3 class="hasAnchor">
<a href="#summarizing-the-estimates-1" class="anchor"></a>Summarizing the estimates:</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(H2AnalysisR)[stat<span class="op">%in%</span><span class="kw">c</span>(<span class="st">"rA"</span>, <span class="st">"H2tMean"</span>, <span class="st">"H2e"</span>, <span class="st">"H2tInf"</span>)]</code></pre></div>
<pre><code>##     method tauQuantile    N   K    stat     MLE PostMean  CI.lower CI.upper tauMean filter
##  1:     PP    V[0%,5%]   28  14      rA  0.3049       NA -0.232088    0.705     3.3    all
##  2:     PP   D[0%,10%]   56  28      rA -0.0180       NA -0.378504    0.350     4.4    all
##  3:     PP   O[0%,12%]   70  35      rA -0.0559       NA -0.375187    0.277     4.9    all
##  4:     PP  qu[0%,20%]  112  56      rA -0.0227       NA -0.280461    0.239     6.1    all
##  5:     PP   Q[0%,25%]  138  69      rA -0.0872       NA -0.314903    0.150     6.7    all
##  6:     PP   M[0%,50%]  278 139      rA  0.0106       NA -0.155350    0.176    10.7    all
##  7:     PP  A[0%,100%]  552 276      rA  0.0283       NA -0.089757    0.146    19.5    all
##  8:  POUMM          NA 1000  NA H2tMean  0.0000   0.0658  0.000065    0.212      NA     NA
##  9:  POUMM          NA 1000  NA     H2e  0.0019   0.0597 -0.075000    0.231      NA     NA
## 10:  POUMM          NA 1000  NA  H2tInf  0.0000   0.0664  0.000065    0.212      NA     NA
## 11:    PMM          NA 1000  NA H2tMean  0.0000   0.0167  0.000020    0.054      NA     NA
## 12:    PMM          NA 1000  NA     H2e  0.0010   0.0051 -0.077709    0.094      NA     NA
## 13:    PMM          NA 1000  NA  H2tInf     NaN   1.0000  1.000000    1.000      NA     NA</code></pre>
</div>
<div id="correlation-profile" class="section level3">
<h3 class="hasAnchor">
<a href="#correlation-profile" class="anchor"></a>Correlation profile</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">corrPrR &lt;-<span class="st"> </span><span class="kw"><a href="../reference/corrProfile.html">corrProfile</a></span>(H2AnalysisR)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotCorrProfile.html">plotCorrProfile</a></span>(H2AnalysisR, corrPrR)</code></pre></div>
<p><img src="UserGuide_files/figure-html/unnamed-chunk-16-1.png" width="652.8"></p>
</div>
<div id="comparing-to-the-true-value-in-the-randomized-data" class="section level3">
<h3 class="hasAnchor">
<a href="#comparing-to-the-true-value-in-the-randomized-data" class="anchor"></a>Comparing to the true value in the randomized data</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pop[, z<span class="op">:</span><span class="er">=</span>z2]

<span class="co"># should be insignificant</span>
<span class="kw"><a href="../reference/R2adj.html">R2adj</a></span>(<span class="dt">data=</span>pop, <span class="dt">activeOnly =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>## [1] -0.0017</code></pre>
</div>
</div>
<div id="packages-used" class="section level1">
<h1 class="hasAnchor">
<a href="#packages-used" class="anchor"></a>Packages used</h1>
<p>Apart from base R functionality, the patherit package uses a number of 3rd party R-packages:</p>
<ul>
<li>For tree processing: ape v5.0 <span class="citation">(Paradis et al. 2016)</span>;</li>
<li>For reporting: data.table v1.10.4.3 <span class="citation">(Dowle and Srinivasan 2016)</span>;</li>
<li>For the POUMM and PP estimates: POUMM v2.0.0 <span class="citation">(Mitov and Stadler 2017)</span>, boot v1.3.18 <span class="citation">(Canty and Ripley 2016)</span>;</li>
<li>For testing: testthat v1.0.2 <span class="citation">(Wickham 2016)</span>.</li>
</ul>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-R-boot">
<p>Canty, Angelo, and Brian Ripley. 2016. <em>Boot: Bootstrap R (S-Plus) Functions.</em> <em>R-Package</em>. <a href="https://cran.r-project.org/web/packages/boot/index.html" class="uri">https://cran.r-project.org/web/packages/boot/index.html</a>.</p>
</div>
<div id="ref-R-data-table">
<p>Dowle, Matt, and Arun Srinivasan. 2016. <em>Data.table: Extension of ‘Data.frame‘</em>. <a href="https://CRAN.R-project.org/package=data.table" class="uri">https://CRAN.R-project.org/package=data.table</a>.</p>
</div>
<div id="ref-Mitov:2016kd">
<p>Mitov, Venelin, and Tanja Stadler. 2016. “The heritability of pathogen traits - definitions and estimators.” <em>Unpublished Data</em> https://www.biorxiv.org/content/early/2016/06/12/058503.</p>
</div>
<div id="ref-R-POUMM">
<p>———. 2017. <em>Fast and Robust Inference of Phylogenetic Ornstein-Uhlenbeck Models Using Parallel Likelihood Calculation</em>. <em>BiorXiv</em>. <a href="http://biorxiv.org/content/early/2017/05/30/115089" class="uri">http://biorxiv.org/content/early/2017/05/30/115089</a>.</p>
</div>
<div id="ref-R-ape">
<p>Paradis, Emmanuel, Simon Blomberg, Ben Bolker, Julien Claude, Hoa Sien Cuong, Richard Desper, Gilles Didier, et al. 2016. <em>Ape: Analyses of Phylogenetics and Evolution</em>. <a href="https://CRAN.R-project.org/package=ape" class="uri">https://CRAN.R-project.org/package=ape</a>.</p>
</div>
<div id="ref-R-testthat">
<p>Wickham, Hadley. 2016. <em>Testthat: Unit Testing for R</em>. <a href="https://CRAN.R-project.org/package=testthat" class="uri">https://CRAN.R-project.org/package=testthat</a>.</p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#extracting-phylogenetic-and-phenotypic-data-from-the-simulated-epidemic">Extracting phylogenetic and phenotypic data from the simulated epidemic</a></li>
      <li><a href="#visualizing-the-data">Visualizing the data</a></li>
      <li><a href="#estimating-the-heritability-using-anova-cpp-poumm-and-the-pmm-methods-">Estimating the heritability using ANOVA-CPP, POUMM and the PMM methods.</a></li>
      <li><a href="#summarizing-the-estimates">Summarizing the estimates:</a></li>
      <li><a href="#generating-a-correlation-profile">Generating a correlation profile</a></li>
      <li><a href="#comparing-to-the-true-value">Comparing to the true value</a></li>
      <li><a href="#randomization-test">Randomization test</a></li>
      <li><a href="#packages-used">Packages used</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Venelin Mitov.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
